### PR TITLE
Make cc_process_args() only use local variables

### PR DESCRIPTION
**Pull request by totph**
plus

Simplify make_relative_path() argument and result ownership

because all the make_relative_path(x_strdup(..)) + free() calls were getting out of hand.

### CODE REVIEW

File: src/ccache.cpp [hunk 1]
```diff
@@ -1730,7 +1568,7 @@ hash_nvcc_host_compiler(struct hash* hash,
 
 // Update a hash with information common for the direct and preprocessor modes.
 static void
-hash_common_info(struct args* args, struct hash* hash)
+hash_common_info(struct args* args, struct hash* hash, const ArgsInfo& ccargs)
```
Comment (jrosdahl) [inline src/ccache.cpp]: Perhaps rename ccargs to args_info?

File: src/ccache.cpp [hunk 2]
```diff
@@ -3190,191 +3038,184 @@ cc_process_args(struct args* args,
...
+
+      result.output_obj = Util::change_extension(result.output_obj, extension);
     }
   }
 
-  if (seen_split_dwarf) {
-    char* p = strrchr(output_obj, '.');
-    if (!p || !p[1]) {
+  if (result.seen_split_dwarf) {
+    size_t pos = result.output_obj.find('.');
```
Comment (jrosdahl) [inline src/ccache.cpp]: This doesn't look correct; `strrchr` corresponds to `.rfind`.

File: src/ccache.cpp [hunk 3]
```diff
@@ -3847,10 +3675,40 @@ ccache(int argc, char* argv[])
   // Arguments to send to the real compiler.
   struct args* compiler_args;
   MTR_BEGIN("main", "process_args");
-  if (!cc_process_args(
-        orig_args, &preprocessor_args, &extra_args_to_hash, &compiler_args)) {
+  ArgsInfo ccargs = cc_process_args(g_config,
```
Comment (jrosdahl) [inline src/ccache.cpp]: Rename to `args_info`?

File: src/legacy_util.cpp [hunk 4]
```diff
@@ -780,6 +780,29 @@ format_parsable_size_with_suffix(uint64_t size)
...
 }
 
+// Convert an owned char* string `str` to an std::string and free `str`.
+std::string
+from_owned_cstr(char* str)
+{
+  std::string result;
+  if (str) {
+    result = str;
+    delete (str);
```
Comment (jrosdahl) [inline src/legacy_util.cpp]:
```suggestion
    delete str;
```

File: src/legacy_util.cpp [hunk 5]
```diff
@@ -780,6 +780,29 @@ format_parsable_size_with_suffix(uint64_t size)
...
+
+// Convert a char* string `str` to a std::string, if `str` is NULL return "".
+std::string
+from_cstr(const char* str)
+{
+  std::string result;
+  if (str) {
+    result = str;
+  }
+  return result;
```
Comment (jrosdahl) [inline src/legacy_util.cpp]:
```suggestion
  return str ? str : {};
```

**Comment by totph** [comment]
Extracted and tested the `from_cstr` functions and added a ScopeGuard to make returning bool easier. All other issues are addressed.

File: src/ccache.cpp [hunk 6]
```diff
@@ -1168,9 +1016,8 @@ result_name_from_depfile(const char* depfile, struct hash* hash)
       if (!has_absolute_include_headers) {
         has_absolute_include_headers = is_absolute_path(token);
       }
-      char* path = make_relative_path(x_strdup(token));
+      std::string path = make_relative_path(token).c_str();
```
Comment (AlexanderLanin) [inline src/ccache.cpp]:
```suggestion
      std::string path = make_relative_path(token);
```

File: src/ccache.cpp [hunk 7]
```diff
@@ -2766,95 +2621,88 @@ cc_process_args(struct args* args,
...
-      profile_arcs = true;
-      generating_coverage = true;
+      args_info.profile_arcs = true;
+      args_info.generating_coverage = true;
       args_add(common_args, argv[i]);
       continue;
     }
     if (str_startswith(argv[i], "-fprofile-dir=")) {
-      profile_dir = x_strdup(argv[i] + 14);
+      args_info.profile_dir = from_owned_cstr(x_strdup(argv[i] + 14));
```
Comment (AlexanderLanin) [inline src/ccache.cpp]: Isn't this the same?
```suggestion
      args_info.profile_dir = from_cstr(argv[i] + 14);
```

**Comment by jrosdahl** [comment]
Thanks (and thanks @AlexanderLanin for the review)!

**Pull request merged**